Add a howto document for transaction fees (Fee 2.0) (#27)
icon-project/documentation: howto/how-to-set-the-policy-for-the-transaction-fee.md

---
title: "How to set the policy for the transaction fees"
excerpt: ""
---

## Overview

> Users must pay the transaction fees to execute a smart contract (SCORE) on the ICON Network. But the fact that all DApp users were required to pay transaction fees has been a major hurdle to attract more users into the DApp service. With the Fee 2.0 feature (Fee Sharing and Virtual Step), the DApp service operators will be able to pay the transaction fees on behalf of the service users.

Reviewer: ```suggestion
Transaction senders must pay the transaction fees to execute a smart contract on the blockchain. The result of this transaction fee system was that all dApp users were required to pay transaction fees to use the service. This has been a major hurdle in attracting more users into the dApp service. With the ICON Transaction Fee 2.0 (Fee Sharing and Virtual Step), dApp service operators can choose to pay the transaction fees on behalf of the service users.
```

Reviewer 2: I think it would be better to use the terminology 'DApp' or 'dApp' in a consistent way. In this comment 'DApp' changed into 'dApp' but I found that the 'DApp' used on following description.(Line 14 of this docs)

This document shows how to use the Fee Sharing and Virtual Step, and what benefits the service operators can get from the new features.

## Intended Audience

> ICON DApp service operators who want the users to use their services without paying any transaction fees.

Author: ```suggestion
ICON dApp service operators who want the users to use their services without paying any transaction fees.
```

## Purpose

Service operators will be able to set the policy for the transaction fees after following this document.

* Service operators can pay the transaction fees on behalf of the service users.
* Service operators can offset the fee burden by using the Virtual Step system.

## Prerequisites

> * Basic concept of the transaction fees and Step on ICON Network

Reviewer: ```suggestion
* Basic concept of the [transaction fees and Step on ICON Network](https://www.icondev.io/docs/transaction-fees)
```

> * Knowledge of the ICON JSON-RPC APIs

Reviewer: ```suggestion
* Knowledge of the [ICON JSON-RPC APIs](https://www.icondev.io/docs/icon-json-rpc-v3)
```

> * Writing SCORE

Reviewer: ```suggestion
* [Writing SCORE](https://www.icondev.io/docs/writing-score)
```

> ## Make a SCORE which supports Fee Sharing

Reviewer: ```suggestion
## Make your SCORE support Fee Sharing
```

> By default, all the transaction fees to execute a SCORE will be charged to the transaction sender. However, the SCORE can determine who will pay the transaction fees during its execution. To do this, we provide the following APIs that can be invoked from the SCORE external method.

Reviewer: ```suggestion
By default, all the transaction fees to execute a SCORE will be charged to the transaction sender. With ICON Fee2.0, the SCORE can determine who will pay the transaction fees during its execution. To do this, `IconScoreBase` provides the following APIs that can be invoked inside the SCORE's external methods that cause a state transition.
```

* `get_fee_sharing_proportion(self) -> int`
    * Returns the current fee sharing proportion of the SCORE.
    * Return value
        * the current fee sharing proportion that the SCORE will pay
        * 100 means the SCORE will pay 100% of transaction fees on behalf of the transaction sender.
        * `int` value between 0 to 100
* `set_fee_sharing_proportion(self, proportion: int)`
    * Sets the proportion of the transaction fees that the SCORE will pay

>     * `proportion` should be between 0 to 100

Reviewer: if the input value is outside the range, be silent or raise an exception?

Reviewer 2: According to Line 75. when the input is not in proper boundary, it will raise exception I guess.

Author: Exception will be raised even if the SCORE didn't check the range correctly.

    * If this method is invoked multiple times, the last proportion value will be used.

### SCORE Example

> 

Reviewer: ```suggestion
Below example SCORE pays the full or partial transaction fee that is required to invoke the `setValue` method for the registered users.
```

> The following example sets the fee sharing proportion according to the white list that can be set separately. If the transaction sender (`self.tx.origin`) is in the white list, some or all of the transaction fees will be charged to the SCORE as much as the proportion value set in the white list.

Reviewer: ```suggestion
The following example sets the fee sharing proportion according to the whitelist that can be set separately. If the transaction sender (`self.tx.origin`) is in the whitelist, part of the transaction fees will be charged to the SCORE as per the proportion value set in the whitelist.
```

        self._whitelist = DictDB("whitelist", db, int)
        self._value = VarDB("value", db, str)

    def on_install(self):
        super().on_install()

        self._value.set("No value")

    def on_update(self):
        super().on_update()

    def _check_owner(self):
        if self.tx.origin != self.owner:
            revert("Invalid SCORE owner")

    @staticmethod
    def _check_proportion(proportion: int):
        if not (0 <= proportion <= 100):
            revert(f"Invalid proportion: {proportion}")

    @external(readonly=True)
    def getProportion(self, address: Address) -> int:
        return self._whitelist[address]

    @external
    def addToWhitelist(self, address: Address, proportion: int = 100):
        self._check_owner()
        self._check_proportion(proportion)

        self._whitelist[address] = proportion

    @external(readonly=True)
    def getValue(self) -> str:
        return self._value.get()

    @external
    def setValue(self, value: str):
        self._value.set(value)

        proportion: int = self._whitelist[self.tx.origin]
        self.set_fee_sharing_proportion(proportion)
        self.ValueSet(self.tx.origin, proportion)
```

> Note that `set_fee_sharing_proportion()` method should be called from the SCORE method (`setValue` in the example above) for the SCORE to pay the transaction fees. Otherwise, the transaction sender will pay all of the transaction fees as before.

Reviewer: ```suggestion
Note that `set_fee_sharing_proportion()` method should be called inside the SCORE method (`setValue` in the example above) that the SCORE is to pay the transaction fee for the transaction sender. Otherwise, the transaction sender will pay all of the transaction fees.
```

Also the transaction sender must have a minimum ICX balance to send the transaction, since it cannot be known beforehand whether the SCORE will pay the transaction fees without executing the transaction actually. However, the users balance will not be changed if the SCORE pays all of the transaction fees.

## Add deposit to the SCORE

> After deploying the example SCORE above to the ICON Network, you need to deposit some ICX to the SCORE for the Fee Sharing and this deposit action will generate Virtual Steps.

Reviewer: ```suggestion
After deploying the example SCORE above to the ICON Network, you need to deposit ICX to the SCORE for the Fee Sharing and this deposit action will generate Virtual Steps.
```

> Currently only the SCORE owner (who deploys the SCORE) can deposit ICX to the SCORE.

Reviewer: ```suggestion
Currently, only the SCORE owner (who deployed the SCORE) can deposit ICX to the SCORE.
```

> To add deposit to the SCORE, you can use the following JSON-RPC API (you can also use ICON SDKs like [Java SDK], [Python SDK], etc).

Reviewer: ```suggestion
To add a deposit to the SCORE, use the following JSON-RPC API or the equivalent in ICON SDKs. See [Java SDK] and [Python SDK].
```

        "stepUsed": "0x1ba94",
        "stepPrice": "0x2540be400",
        "cumulativeStepUsed": "0x1ba94",
        "eventLogs": [
            {
                "scoreAddress": "cx216e1468b780ac1b54c328d19ea23a35a6899e55",
                "indexed": [
                    "DepositAdded(bytes,Address,int,int)",
                    "0x64b118d4a3c2b3b93362a0f3ea06e5519de42449523465265b85509041e69011",
                    "hxe7af5fcfd8dfc67530a01a0e403882687528dfcb"
                ],
                "data": [
                    "0x10f0cf064dd59200000",
                    "0x13c680"
                ]
            }
        ],
        "logsBloom": "0x00000001000000000000000000000...0000000000000000010000000000000",
        "status": "0x1"
    },
    "id": 1
}
```

The declaration of `DepositAdded` eventlog is as follows.

```python
@eventlog(indexed=2)
def DepositAdded(self, id: bytes, from_: Address, amount: int, term: int):
    pass
```
`id` is the deposit id (i.e., the transaction hash), `from_` is the transaction sender, `amount` is the amount of deposited ICX, and `term` is the deposit period in blocks, which is currently fixed to 1 month (1,296,000 blocks in 30 days).

> ## Check the fee deduction after calling SCORE methods

Reviewer: ```suggestion
## Invoke SCORE method and check the fee deduction result
```

Now you have a SCORE that has the ICX deposit.

> You need to add some users to the white list first before letting them to call the SCORE method.

Reviewer: ```suggestion
Before sending a transaction, you need to add the user to the whitelist.
```

> After that, if a user that is in the white list calls the SCORE method (`setValue` in this case), some or all of the transaction fees will be charged to the SCORE.

Reviewer: ```suggestion
If a user who is in the whitelist calls the SCORE method (`setValue` in this case), part of the transaction fee will be charged to the SCORE according to the proportion set in the whitelist.
```

```json
{
    "jsonrpc": "2.0",
    "result": {
        "txHash": "0x5ca11a63e24cdb837026d4a20165d0246ecfc7bc40d84eaa33bdf43538b318c8",
        "blockHeight": "0x17",
        "blockHash": "0x53e6c94078e6b94135fe773675493dcc47ea62e8878e48eacd3044f5068d7bf1",
        "txIndex": "0x1",
        "to": "cx216e1468b780ac1b54c328d19ea23a35a6899e55",
        "stepUsed": "0x21bc4",
        "stepPrice": "0x2540be400",
        "cumulativeStepUsed": "0x43788",
        "eventLogs": [
            {
                "scoreAddress": "cx216e1468b780ac1b54c328d19ea23a35a6899e55",
                "indexed": [
                    "ValueSet(Address,int)",
                    "hx93a1562d85982de882b5fa4df05d42d35a7db0b1"
                ],
                "data": [
                    "0x32"
                ]
            }
        ],
        "logsBloom": "0x00000001000000000000000000000...0000000000000000010000000000000",
        "status": "0x1",
        "stepUsedDetails": {
            "hx93a1562d85982de882b5fa4df05d42d35a7db0b1": "0x10de2",
            "cx216e1468b780ac1b54c328d19ea23a35a6899e55": "0x10de2"
        }
    },
    "id": 1
}
```

> Note that the transaction result will be the same as when there was no Fee 2.0 feature, if either the user is not in the white list or the SCORE does not set the proportion (i.e., there is no `stepUsedDetails` in the transaction result).

Reviewer: ```suggestion
Note that the transaction result will be the same as when there was no Fee 2.0 feature if either the user is not in the whitelist or the SCORE does not set the fee sharing proportion (i.e., there will be no `stepUsedDetails` in the transaction result).
```
